clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -2857,32 +2857,63 @@ static void handleWeakImportAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
 }
 
 // Checks correctness of mutual usage of different work_group_size attributes:
-// reqd_work_group_size, max_work_group_size. Values of reqd_work_group_size
-// arguments shall be equal or less than values coming from max_work_group_size.
+// reqd_work_group_size, max_work_group_size and max_global_work_dim.
+// Values of reqd_work_group_size arguments shall be equal or less than values
+// coming from max_work_group_size.
+// In case of 'max_global_work_dim' attribute equals to 0 we shall ensure
+// that if max_work_group_size and reqd_work_group_size attributes exist,
+// then they are equal (1, 1, 1).
 static bool checkWorkGroupSizeValues(Sema &S, Decl *D, const ParsedAttr &Attr,
```

Comment:
I have a major concern with how we're using this function. In particular, this routine:

1. handling of each `WorkGroupAttrT` attribute results in a call to `checkWorkGroupSizeValues()`;
2. inside `checkWorkGroupSizeValues()`, we lookup every related `WGAttrT` attribute for the second time.

If my subsequent cringe is baseless, please challenge it:
Imagine a normal use-case with `max_global_work_dim`, `max_work_group_size`, `reqd_work_group_size` being applied to the same kernel. It looks like we're going to call `D->getAttr<WGAttrT>` 9 times, 3 times per each attribute.

Myself, I haven't worked out a way out of this yet. In case this whole concern **is** substantial from your point of view, could you please join this "quest for a better solution"?

Also, @erichkeane, could you please chime in?

Reply:
But all of these calls are done for different purposes: 1. check, if attribute is duplicated on function decl 2. check values correctness depending on other attributes applied. And, actually, I haven't found a third call. Will try tomorrow, since it's 2 a.m. right now :)

Reply:
Could we somehow move the call to `checkWorkGroupSizeValues` into `Sema::ProcessDeclAttributes()` so that its result is known before we enter the vast switch over `ParsedAttr`s? This way, we can run the attributes compatibility check only once, and not even bother ourselves with calls to `handle<WorkGroupAttrT>()` if the result is `false`.

Reply:
Point here, that `checkWorkGroupSizeValues` is called once per `ProcessDeclAttribute` function and therefore for the first run on the decl - only one attr is parsed, on the second - two attrs are parsed and so on.